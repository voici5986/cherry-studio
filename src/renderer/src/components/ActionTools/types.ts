/**
 * 动作工具基本信息
 */
export interface ActionToolSpec {
  id: string
  type: 'core' | 'quick'
  order: number
}

/**
 * 动作工具定义接口
 * @param id 唯一标识符
 * @param type 工具类型
 * @param order 显示顺序，越小越靠右
 * @param icon 按钮图标
 * @param tooltip 提示文本
 * @param visible 显示条件
 * @param onClick 点击动作
 * @param children 子工具（例如 more 下拉菜单）
 */
export interface ActionTool extends ActionToolSpec {
  icon: React.ReactNode
  tooltip?: string
  visible?: () => boolean
  onClick?: () => void
  children?: Omit<ActionTool, 'children'>[]
}

/**
 * 子组件向父组件注册工具所需的 props
 */
export interface ToolRegisterProps {
  setTools?: (value: React.SetStateAction<ActionTool[]>) => void
}
