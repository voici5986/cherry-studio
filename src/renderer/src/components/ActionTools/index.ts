export * from './constants'
export * from './hooks/useImageTools'
export * from './hooks/useToolManager'
export * from './types'
