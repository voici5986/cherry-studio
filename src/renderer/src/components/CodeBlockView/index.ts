export { default as HtmlArtifactsCard } from './HtmlArtifactsCard'
export * from './types'
export * from './view'
