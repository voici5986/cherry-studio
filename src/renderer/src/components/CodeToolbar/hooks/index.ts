export * from './useCopyTool'
export * from './useDownloadTool'
export * from './useExpandTool'
export * from './useRunTool'
export * from './useSaveTool'
export * from './useSplitViewTool'
export * from './useViewSourceTool'
export * from './useWrapTool'
