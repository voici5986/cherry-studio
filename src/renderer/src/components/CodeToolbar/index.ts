export { default as CodeToolButton } from './button'
export * from './hooks'
export { default as CodeToolbar } from './toolbar'
