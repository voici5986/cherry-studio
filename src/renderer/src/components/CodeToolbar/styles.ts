import styled from 'styled-components'

export const ToolWrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 4px;
  cursor: pointer;
  user-select: none;
  transition: all 0.2s ease;
  color: var(--color-text-3);

  &:hover {
    background-color: var(--color-background-soft);
    .tool-icon {
      color: var(--color-text-1);
    }
  }

  &.active {
    color: var(--color-primary);
    .tool-icon {
      color: var(--color-primary);
    }
  }

  /* For Lucide icons */
  .tool-icon {
    width: 14px;
    height: 14px;
    color: var(--color-text-3);
  }
`
