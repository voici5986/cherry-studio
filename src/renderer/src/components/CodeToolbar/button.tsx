import { ActionTool } from '@renderer/components/ActionTools'
import { Dropdown, Tooltip } from 'antd'
import { memo, useMemo } from 'react'

import { ToolWrapper } from './styles'

interface CodeToolButtonProps {
  tool: ActionTool
}

const CodeToolButton = ({ tool }: CodeToolButtonProps) => {
  const mainTool = useMemo(
    () => (
      <Tooltip key={tool.id} title={tool.tooltip} mouseEnterDelay={0.5} mouseLeaveDelay={0}>
        <ToolWrapper onClick={tool.onClick}>{tool.icon}</ToolWrapper>
      </Tooltip>
    ),
    [tool]
  )

  if (tool.children?.length && tool.children.length > 0) {
    return (
      <Dropdown
        menu={{
          items: tool.children.map((child) => ({
            key: child.id,
            label: child.tooltip,
            icon: child.icon,
            onClick: child.onClick
          }))
        }}
        trigger={['click']}>
        {mainTool}
      </Dropdown>
    )
  }

  return mainTool
}

export default memo(CodeToolButton)
