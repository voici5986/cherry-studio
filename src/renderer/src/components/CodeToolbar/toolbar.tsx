import { ActionTool } from '@renderer/components/ActionTools'
import { HStack } from '@renderer/components/Layout'
import { Tooltip } from 'antd'
import { EllipsisVertical } from 'lucide-react'
import { memo, useMemo, useState } from 'react'
import { useTranslation } from 'react-i18next'
import styled from 'styled-components'

import CodeToolButton from './button'
import { ToolWrapper } from './styles'

const CodeToolbar = ({ tools }: { tools: ActionTool[] }) => {
  const [showQuickTools, setShowQuickTools] = useState(false)
  const { t } = useTranslation()

  // 根据条件显示工具
  const visibleTools = tools.filter((tool) => !tool.visible || tool.visible())

  // 按类型分组
  const coreTools = visibleTools.filter((tool) => tool.type === 'core')
  const quickTools = visibleTools.filter((tool) => tool.type === 'quick')

  // 点击了 more 按钮或者只有一个快捷工具时
  const quickToolButtons = useMemo(() => {
    if (quickTools.length === 1 || (quickTools.length > 1 && showQuickTools)) {
      return quickTools.map((tool) => <CodeToolButton key={tool.id} tool={tool} />)
    }

    return null
  }, [quickTools, showQuickTools])

  if (visibleTools.length === 0) {
    return null
  }

  return (
    <StickyWrapper>
      <ToolbarWrapper className="code-toolbar">
        {/* 有多个快捷工具时通过 more 按钮展示 */}
        {quickToolButtons}
        {quickTools.length > 1 && (
          <Tooltip title={t('code_block.more')} mouseEnterDelay={0.5}>
            <ToolWrapper onClick={() => setShowQuickTools(!showQuickTools)} className={showQuickTools ? 'active' : ''}>
              <EllipsisVertical className="tool-icon" />
            </ToolWrapper>
          </Tooltip>
        )}

        {/* 始终显示核心工具 */}
        {coreTools.map((tool) => (
          <CodeToolButton key={tool.id} tool={tool} />
        ))}
      </ToolbarWrapper>
    </StickyWrapper>
  )
}

const StickyWrapper = styled.div`
  position: sticky;
  top: 28px;
  z-index: 10;
`

const ToolbarWrapper = styled(HStack)`
  position: absolute;
  align-items: center;
  bottom: 0.3rem;
  right: 0.5rem;
  height: 24px;
  gap: 4px;
`

export default memo(CodeToolbar)
