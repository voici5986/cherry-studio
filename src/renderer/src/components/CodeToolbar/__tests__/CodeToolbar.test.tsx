import { ActionTool } from '@renderer/components/ActionTools'
import { fireEvent, render, screen } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import CodeToolbar from '../toolbar'

// Test constants
const MORE_BUTTON_TOOLTIP = 'code_block.more'

// Mock components
const mocks = vi.hoisted(() => ({
  CodeToolButton: vi.fn(({ tool }) => (
    <div data-testid={`tool-button-${tool.id}`} data-tool-id={tool.id} data-tool-type={tool.type}>
      {tool.icon}
    </div>
  )),
  Tooltip: vi.fn(({ children, title }) => (
    <div data-testid="tooltip" data-title={title}>
      {children}
    </div>
  )),
  HStack: vi.fn(({ children, className }) => (
    <div data-testid="hstack" className={className}>
      {children}
    </div>
  )),
  ToolWrapper: vi.fn(({ children, onClick, className }) => (
    <div data-testid="tool-wrapper" onClick={onClick} className={className} role="button" tabIndex={0}>
      {children}
    </div>
  )),
  EllipsisVertical: vi.fn(() => <div data-testid="ellipsis-icon" className="tool-icon" />),
  useTranslation: vi.fn(() => ({
    t: vi.fn((key: string) => key)
  }))
}))

vi.mock('../button', () => ({
  default: mocks.CodeToolButton
}))

vi.mock('antd', () => ({
  Tooltip: mocks.Tooltip
}))

vi.mock('@renderer/components/Layout', () => ({
  HStack: mocks.HStack
}))

vi.mock('./styles', () => ({
  ToolWrapper: mocks.ToolWrapper
}))

vi.mock('lucide-react', () => ({
  EllipsisVertical: mocks.EllipsisVertical
}))

vi.mock('react-i18next', () => ({
  useTranslation: mocks.useTranslation
}))

// Helper function to create mock tools
const createMockTool = (overrides: Partial<ActionTool> = {}): ActionTool => ({
  id: 'test-tool',
  type: 'core',
  order: 1,
  icon: <div data-testid="test-icon">Icon</div>,
  tooltip: 'Test Tool',
  onClick: vi.fn(),
  ...overrides
})

// Common test data
const createMixedTools = () => [
  createMockTool({ id: 'quick1', type: 'quick' }),
  createMockTool({ id: 'quick2', type: 'quick' }),
  createMockTool({ id: 'core1', type: 'core' })
]

const createCoreOnlyTools = () => [
  createMockTool({ id: 'core1', type: 'core' }),
  createMockTool({ id: 'core2', type: 'core' })
]

// Helper function to click more button
const clickMoreButton = () => {
  const tooltip = screen.getByTestId('tooltip')
  fireEvent.click(tooltip.firstChild as Element)
}

describe('CodeToolbar', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('basic rendering', () => {
    it('should match snapshot with mixed tools', () => {
      const { container } = render(<CodeToolbar tools={createMixedTools()} />)
      expect(container).toMatchSnapshot()
    })

    it('should match snapshot with only core tools', () => {
      const { container } = render(<CodeToolbar tools={[createMockTool({ id: 'core1', type: 'core' })]} />)
      expect(container).toMatchSnapshot()
    })
  })

  describe('empty state', () => {
    it('should render nothing when no tools provided', () => {
      const { container } = render(<CodeToolbar tools={[]} />)
      expect(container.firstChild).toBeNull()
    })

    it('should render nothing when all tools are not visible', () => {
      const tools = [
        createMockTool({ id: 'tool1', visible: () => false }),
        createMockTool({ id: 'tool2', visible: () => false })
      ]
      const { container } = render(<CodeToolbar tools={tools} />)
      expect(container.firstChild).toBeNull()
    })
  })

  describe('tool visibility filtering', () => {
    it('should only render visible tools', () => {
      const tools = [
        createMockTool({ id: 'visible-tool', visible: () => true }),
        createMockTool({ id: 'hidden-tool', visible: () => false }),
        createMockTool({ id: 'no-visible-prop' }) // Should be visible by default
      ]
      render(<CodeToolbar tools={tools} />)

      expect(screen.getByTestId('tool-button-visible-tool')).toBeInTheDocument()
      expect(screen.getByTestId('tool-button-no-visible-prop')).toBeInTheDocument()
      expect(screen.queryByTestId('tool-button-hidden-tool')).not.toBeInTheDocument()
    })

    it('should show tools without visible function by default', () => {
      const tools = [createMockTool({ id: 'default-visible' })]
      render(<CodeToolbar tools={tools} />)

      expect(screen.getByTestId('tool-button-default-visible')).toBeInTheDocument()
    })
  })

  describe('tool type grouping and quick tools behavior', () => {
    it('should separate core and quick tools - show quick tools when expanded', () => {
      const tools = [
        createMockTool({ id: 'core1', type: 'core' }),
        createMockTool({ id: 'quick1', type: 'quick' }),
        createMockTool({ id: 'core2', type: 'core' }),
        createMockTool({ id: 'quick2', type: 'quick' })
      ]
      render(<CodeToolbar tools={tools} />)

      // Initial state: core tools visible, quick tools hidden
      expect(screen.getByTestId('tool-button-core1')).toBeInTheDocument()
      expect(screen.getByTestId('tool-button-core2')).toBeInTheDocument()
      expect(screen.queryByTestId('tool-button-quick1')).not.toBeInTheDocument()
      expect(screen.queryByTestId('tool-button-quick2')).not.toBeInTheDocument()

      // After clicking more button, quick tools should be visible
      clickMoreButton()

      expect(screen.getByTestId('tool-button-quick1')).toBeInTheDocument()
      expect(screen.getByTestId('tool-button-quick2')).toBeInTheDocument()
    })

    it('should render only core tools when no quick tools exist', () => {
      render(<CodeToolbar tools={createCoreOnlyTools()} />)

      expect(screen.getByTestId('tool-button-core1')).toBeInTheDocument()
      expect(screen.getByTestId('tool-button-core2')).toBeInTheDocument()
      expect(screen.queryByTestId('tooltip')).not.toBeInTheDocument() // No more button
    })

    it('should show single quick tool directly without more button', () => {
      const tools = [createMockTool({ id: 'quick1', type: 'quick' }), createMockTool({ id: 'core1', type: 'core' })]
      render(<CodeToolbar tools={tools} />)

      expect(screen.getByTestId('tool-button-quick1')).toBeInTheDocument()
      expect(screen.getByTestId('tool-button-core1')).toBeInTheDocument()
      expect(screen.queryByTestId('tooltip')).not.toBeInTheDocument() // No more button
    })

    it('should show more button when multiple quick tools exist', () => {
      render(<CodeToolbar tools={createMixedTools()} />)

      // Initially quick tools should be hidden
      expect(screen.queryByTestId('tool-button-quick1')).not.toBeInTheDocument()
      expect(screen.queryByTestId('tool-button-quick2')).not.toBeInTheDocument()
      expect(screen.getByTestId('tool-button-core1')).toBeInTheDocument()
      expect(screen.getByTestId('tooltip')).toBeInTheDocument() // More button exists
    })

    it('should toggle quick tools visibility when more button is clicked', () => {
      render(<CodeToolbar tools={createMixedTools()} />)

      // Initial state: quick tools hidden
      expect(screen.queryByTestId('tool-button-quick1')).not.toBeInTheDocument()
      expect(screen.queryByTestId('tool-button-quick2')).not.toBeInTheDocument()

      // Click more button: quick tools visible
      clickMoreButton()
      expect(screen.getByTestId('tool-button-quick1')).toBeInTheDocument()
      expect(screen.getByTestId('tool-button-quick2')).toBeInTheDocument()

      // Click more button again: quick tools hidden
      clickMoreButton()
      expect(screen.queryByTestId('tool-button-quick1')).not.toBeInTheDocument()
      expect(screen.queryByTestId('tool-button-quick2')).not.toBeInTheDocument()
    })

    it('should apply active class to more button when quick tools are shown', () => {
      const tools = [createMockTool({ id: 'quick1', type: 'quick' }), createMockTool({ id: 'quick2', type: 'quick' })]
      render(<CodeToolbar tools={tools} />)

      const tooltip = screen.getByTestId('tooltip')
      const moreButton = tooltip.firstChild as Element

      // Initial state: no active class
      expect(moreButton).not.toHaveClass('active')

      // After click: has active class
      fireEvent.click(moreButton)
      expect(moreButton).toHaveClass('active')

      // After second click: no active class
      fireEvent.click(moreButton)
      expect(moreButton).not.toHaveClass('active')
    })

    it('should display correct tooltip and icon for more button', () => {
      render(<CodeToolbar tools={createMixedTools()} />)

      const tooltip = screen.getByTestId('tooltip')
      expect(tooltip).toHaveAttribute('data-title', MORE_BUTTON_TOOLTIP)

      expect(screen.getByTestId('ellipsis-icon')).toBeInTheDocument()
      expect(screen.getByTestId('ellipsis-icon')).toHaveClass('tool-icon')
    })

    it('should render core tools regardless of quick tools state', () => {
      const tools = [
        createMockTool({ id: 'quick1', type: 'quick' }),
        createMockTool({ id: 'quick2', type: 'quick' }),
        createMockTool({ id: 'core1', type: 'core' }),
        createMockTool({ id: 'core2', type: 'core' })
      ]
      render(<CodeToolbar tools={tools} />)

      // Core tools always visible
      expect(screen.getByTestId('tool-button-core1')).toBeInTheDocument()
      expect(screen.getByTestId('tool-button-core2')).toBeInTheDocument()

      // After clicking more button, core tools still visible
      clickMoreButton()
      expect(screen.getByTestId('tool-button-core1')).toBeInTheDocument()
      expect(screen.getByTestId('tool-button-core2')).toBeInTheDocument()
    })
  })
})
