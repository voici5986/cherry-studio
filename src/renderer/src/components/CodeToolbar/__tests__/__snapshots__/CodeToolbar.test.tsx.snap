// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CodeToolbar > basic rendering > should match snapshot with mixed tools 1`] = `
.c2 {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 4px;
  cursor: pointer;
  user-select: none;
  transition: all 0.2s ease;
  color: var(--color-text-3);
}

.c2:hover {
  background-color: var(--color-background-soft);
}

.c2:hover .tool-icon {
  color: var(--color-text-1);
}

.c2.active {
  color: var(--color-primary);
}

.c2.active .tool-icon {
  color: var(--color-primary);
}

.c2 .tool-icon {
  width: 14px;
  height: 14px;
  color: var(--color-text-3);
}

.c0 {
  position: sticky;
  top: 28px;
  z-index: 10;
}

.c1 {
  position: absolute;
  align-items: center;
  bottom: 0.3rem;
  right: 0.5rem;
  height: 24px;
  gap: 4px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 code-toolbar"
      data-testid="hstack"
    >
      <div
        data-testid="tooltip"
        data-title="code_block.more"
      >
        <div
          class="c2"
        >
          <div
            class="tool-icon"
            data-testid="ellipsis-icon"
          />
        </div>
      </div>
      <div
        data-testid="tool-button-core1"
        data-tool-id="core1"
        data-tool-type="core"
      >
        <div
          data-testid="test-icon"
        >
          Icon
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`CodeToolbar > basic rendering > should match snapshot with only core tools 1`] = `
.c0 {
  position: sticky;
  top: 28px;
  z-index: 10;
}

.c1 {
  position: absolute;
  align-items: center;
  bottom: 0.3rem;
  right: 0.5rem;
  height: 24px;
  gap: 4px;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1 code-toolbar"
      data-testid="hstack"
    >
      <div
        data-testid="tool-button-core1"
        data-tool-id="core1"
        data-tool-type="core"
      >
        <div
          data-testid="test-icon"
        >
          Icon
        </div>
      </div>
    </div>
  </div>
</div>
`;
